src/esp/gfx/PTexMeshDrawable.cpp

```diff
@@ -18,16 +18,19 @@ PTexMeshDrawable::PTexMeshDrawable(
     : Drawable{node, shader, ptexMeshData.getRenderingBuffer(submeshID)->mesh,
                group},
       tex_(ptexMeshData.getRenderingBuffer(submeshID)->tex),
-      adjTex_(ptexMeshData.getRenderingBuffer(submeshID)->adjTex),
+      adjFaces_(ptexMeshData.getRenderingBuffer(submeshID)->adjFaces),
       tileSize_(ptexMeshData.tileSize()),
-      exposure_(ptexMeshData.exposure()) {}
+      exposure_(ptexMeshData.exposure()),
+      gamma_(ptexMeshData.gamma()),
+      saturation_(ptexMeshData.saturation()) {}
 
 void PTexMeshDrawable::draw(const Magnum::Matrix4& transformationMatrix,
                             Magnum::SceneGraph::Camera3D& camera) {
-  adjTex_.bind(1);
+  adjFaces_.bind(1);
```

Comment:
@mosra : adjFaces is the buffer texture, and it is bound to 1. Not sure if 1 is correct binding point.

Reply:
I don't know -- I don't see any `layout(binding = N)`  on the shader side, or anything related in the constructor, for either this texture or the other, which leads me to think that this is the root of the problem. To be clear -- Magnum doesn't have any implicit binding points or anything, so you can't expect the shader to pick (samplers, uniforms, attributes) up unless you explicitly define the location / binding points for these as well. 

*(This comment is getting a bit long, so bear with me please.)*

What's the designated way of doing this in Magnum (and what you are doing here differently) is, step-by-step ([docs for reference](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1AbstractShaderProgram.html#GL-AbstractShaderProgram-subclassing) -- look there for more details about each point):

- For each uniform, defining `layout(location = N)` in the shader code, and then using the same `N` on C++ side. (Extensions needed for that are available on all systems we care about, so no need to save results of `uniformLocation()` on the C++ side.) Especially *not* using `setUniform(uniformLocation("some string"), value)` each time in every frame, all those string operations are extremely wasteful.
- For each input attribute, `typedef`ing it in the header with location corresponding to the `layout()` qualifier in the shader code *and* type matching the type you are most likely to have the vertex data in. GL pads the remaining components with `(0, 0, 0, 1)` so for example, if you pass a `Vector2` and have `vec4` on the shader side, you'll get `(x, y, 0, 1)` there -- no need to supply the `0` and `1` explicitly. Here you're using `Vector4` and I don't see a reason to supply the fourth component, so it could be `Vector3` I think, saving 25% of memory. Of course the `PTexMeshData` needs to be updated as well.
- For each sampler, either doing `layout(binding = N)` in the shader code or calling `setUniform(uniformLocation("sampler name"), N)` in the constructor. Here you need the latter, as macOS doesn't have `GL_ARB_shading_language_420pack`. Then, for each sampler providing `bindSomething()` APIs on the C++ side, using the same `N` as a parameter to `texture.bind()`. Here you have `bindTexture()` just for one of the samplers and it has the binding point as a parameter, which doesn't make sense, and in the other case you're calling `bind()` manually, with the `1` having no relation to the shader at all. What I'm usually doing is defining an enum with the binding point IDs and then using those enums for both the `bind()` and `setUniform()` calls -- so for example:

  ```cpp
  enum: UnsignedInt {
    AtlasTextureUnit = 0,
    AdjacencyTextureUnit = 1
  };

  ...

  PTexMeshShader::PTexMeshShader() {
    setUniform(uniformLocation("atlasTex"), AtlasTextureUnit);
    setUniform(uniformLocation("meshAdjFaces"), AdjacencyTextureUnit);
  }

  PTexMeshDhader& PTexMeshShader::bindAtlasTexture(GL::Texture2D& texture) {
    texture.bind(AtlasTextureUnit);
    return *this;
  }

  PTexMeshDhader& PTexMeshShader::bindAdjacencyTexture(GL::BufferTexture& texture) {
    texture.bind(AdjacencyTextureUnit);
    return *this;
  }
  ```

  This way the shader API on the C++ side is complete and you can be sure that you are always binding the right type of a texture to the right binding point. It's generally advised to choose mutually exclusive texture units for each shader (unless it's expected to use the same textures in each), as you save on texture rebinding when switching between different shaders. So e.g. if the generic shader makes use of units 0-3, the PTex shader would take units 4 and 5.

Reply:
Thank you so much for the informative and invaluable comment! It saves me from a lot of guesswork on Magnum usage.

Reply:
> But "Explicit uniform location" is not supported until 4.3.

Right, sorry. For some reason I thought it got in much earlier.

> It makes things complicated. But do you think it is necessary or it can reduce the overhead?

Uh oh, i meant something else -- the `uniformLocation()`, due to all its string operations, is the problem, not `setUniform()`. So call it once in the constructor, save the ID and then do just `setUniform(exposureUniform_, value)`. Check the performance plot in #151 -- just by going from `GenericShader` (which calls `uniformLocation()` every time) to magnum's `Shaders::Flat` (which caches the locations), I was able to shave some additional milliseconds.

Long time ago I tried to cache the uniform values as well, but it didn't prove any measurable performance improvement. And the maintenance overhead was absolutely not worth it. What could make sense is going with UBOs, but those are known to be even slower than classic `setUniform()` calls on Mac and some Intel GPUs, so ... ¯\\\_(ツ)_/¯